unified-scheduler-pool/src/lib.rs

```diff
@@ -530,9 +530,68 @@ impl<S: SpawnableScheduler<TH>, TH: TaskHandler> ThreadManager<S, TH> {
     }
 
     fn start_threads(&mut self, context: &SchedulingContext) {
-        let (mut runnable_task_sender, runnable_task_receiver) =
+        // Firstly, setup bi-directional messaging between the scheduler and handlers to pass
+        // around tasks, by creating 2 channels (one for to-be-handled tasks from the scheduler to
+        // the handlers and the other for finished tasks from the handlers to the scheduler).
+        // Furthermore, this pair of channels is duplicated to work as a primitive 2-level priority
+        // queue, totalling 4 channels.
+        //
+        // This quasi-priority-queue arrangement is desired as an optimization to prioritize
+        // blocked tasks.
+        //
+        // As a quick background, SchedulingStateMachine doesn't throttle runnable tasks at all.
+        // Thus, it's likely for to-be-handled tasks to be stalled for extended duration due to
+        // excessive buffering (commonly known as buffer bloat). Normally, this buffering isn't
+        // problematic and actually intentional to fully saturate all the handler threads.
+        //
+        // However, there's one caveat: task dependencies. It can be hinted with tasks being
+        // blocked, that there could be more similarly-blocked tasks in the future. Empirically,
+        // clearing these linearized long runs of blocking tasks out of the buffer is delaying bank
+        // freezing while only using 1 handler thread or two near the end of slot, deteriorating
+        // the overall concurrency.
+        //
+        // To alleviate the situation, blocked tasks are exchanged via independent communication
+        // pathway as a heuristic for expedite processing. Without prioritization of these tasks,
+        // progression of clearing these runs would be severely hampered due to interleaved
+        // not-blocked tasks (called _idle_ here; typically, voting transactions) in the single
+        // buffer.
+        //
+        // Concurrent priority queue isn't used to avoid penalized throughput due to higher
+        // overhead than crossbeam channel, even considering the doubled processing of the
+        // crossbeam channel. Fortunately, just 2-level prioritization is enough. Also, sticking to
+        // crossbeam was convenient and there was no popular and promising crate for concurrent
+        // priority queue as of writing.
```

Comment:
Maybe I'm misreading this, but it seems odd to say first that the overhead is higher for concurrent priority queues, then say there are no currently promising crates.

Which crates were even considered? If there were a good concurrent priority queue, would you still not choose it for above reasons?

Reply:
> Which crates were even considered?

I was hoping someone to create such one, based on `crossbeam_skiplist` (I'm fan of crossbeam). but none is found on the intertube. lol Also, it has still some scary bugs like this: https://github.com/crossbeam-rs/crossbeam/issues/1023. I'd suspect there's little need for concurrent priority queue after all..

Lastly, I don't have enough time to write own. ;)

> If there were a good concurrent priority queue, would you still not choose it for above reasons?

hehe, nice question. I'll write down here not to make the in-source comment any longer...

I would still not choose it because there are four reasons:

1. generally speaking insertion cost of normal mpmc is `O(1)`. on the other hand, the insertion cost of priority queue is `O(log)`. This is a stark theoretical complexity difference here.
2. Also, priority queue incurs out-of-order memory alloc and dealloc by definition. in turn, some kind of gc is needed. crossbeam uses the epoch strategy. but it also incurs additional impl-imposed overhead on top of the above theoretical overhead.
3. fundamentally, there's little merit of introducing reordering here. as banking stage's block packing strategy improves, transactions should tend to be ordered already. So, generally not worth the overhead. task reordering would make sense when put in the very front line of uncontrolled task sources like third party http requests or before sigverfiy in our case.
4. In contrast, when unified scheduler is put inside a banking stage, it makes sense to introduce priority queue here, in terms of maximizing block rewards, assuming the number of concurrently-executable tasks is more than N worker threads. However, it would easily price out vote-txes and breaks the local fee market thing altogether, which is against my egotistic ideology. So, I'd never put such impl.

Reply:
> Lastly, I don't have enough time to write own. ;)

I partially asked for greedy reasons since I have also wanted/needed a concurrent priority queue in the banking stage scheduler and haven't found a suitable crate yet.
Best I've seen thus far is the skiplist (skipset) inside of crossbeam, but it also doesn't seem particularly great to me.

Appreciate the explanation. I am not entirely opposed to the multi-channel idea, but just wanted to understand path taken.